flexget/plugins/clients/aria2.py

```diff
@@ -12,6 +15,135 @@
 logger = logger.bind(name='aria2')
 
 
+class JSONRPC(object):
+
+    RPC_ID = 'FLEXGET'
+    RPC_VERSION = '2.0'
+
+    ADDURI_METHOD = 'aria2.addUri'
+    GETGLOBALSTAT_METHOD = 'aria2.getGlobalStat'
+    ADDTORRENT_METHOD = 'aria2.addTorrent'
+    ADDMETALINK_METHOD = 'aria2.addMetalink'
+
+    def __init__(self, server, port, scheme, rpc_path, username=None, password=None, token=None):
+        if token:
+            self._token = 'token:' + token
+        else:
+            self._token = ''
+        if username and password:
+            userpass = '%s:%s@' % (username, password)
+        else:
+            userpass = ''
+        url = '%s://%s%s:%s/%s' % (scheme, userpass, server, port, rpc_path)
+        self._url = url
+        logger.debug('aria2 url: {}', url)
+        logger.info('Connecting to daemon at {}', url)
+        self.get_global_stat()
+
+    def _get_req_params(self, method, params=[]):
+        req_params = {
+            'jsonrpc': JSONRPC.RPC_VERSION,
+            'id': JSONRPC.RPC_ID,
+            'method': method,
+            'params': params,
+        }
+        if self._token:
+            req_params['params'].insert(0, self._token)
+        if not req_params['params']:
+            del req_params['params']
+        return req_params
+
+    def _default_error_handle(code, message):
+        logger.critical('Fault code {} message {}', code, message)
+        raise plugin.PluginError('Fault code %s message %s' % (code, message), logger)
+
+    def _default_success_handle(response):
+        return response.text
+
+    def _post(
+        self, method, params, on_success=_default_success_handle, on_fail=_default_error_handle
+    ):
+        _params = self._get_req_params(method, params)
+        resp = requests.post(self._url, data=json.dumps(_params))
+        result = resp.json()
+        if "error" in result:
+            return on_fail(result["error"]["code"], result["error"]["message"])
+        else:
+            return on_success(resp)
+
+    def add_uri(self, uris, options):
+        # https://aria2.github.io/manual/en/html/aria2c.html#aria2.addUri
+        return self._post(JSONRPC.ADDURI_METHOD, params=[[uris], options])
+
+    def add_torrent(self, torrent, options):
+        # https://aria2.github.io/manual/en/html/aria2c.html#aria2.addTorrent
+        params = [torrent]
+        return self._post(JSONRPC.ADDTORRENT_METHOD, params=[torrent, options])
+
+    def get_global_stat(self):
+        # https://aria2.github.io/manual/en/html/aria2c.html#aria2.getGlobalStat
+        return self._post(JSONRPC.GETGLOBALSTAT_METHOD, params=[])
+
+
+class XMLRPC(object):
+    def __init__(self, server, port, scheme, rpc_path, username=None, password=None, token=None):
+        if token:
+            self._token = 'token:' + token
+        else:
+            self._token = ''
+        if username and password:
+            userpass = '%s:%s@' % (username, password)
+        else:
+            userpass = ''
```

Comment:
This code is duplicated across both methods. maybe make a helper function to do it?